src/main/java/org/skriptlang/skript/bukkit/recipes/RecipeWrapper.java

```diff
@@ -0,0 +1,385 @@
+		return result;
+	}
+
+	public NamespacedKey getKey() {
+		return this.key;
+	}
+
+	public RecipeType getRecipeType() {
+		return recipeType;
+	}
+
+	public void addError(String error) {
+		errors.add(error);
+	}
+
+	public List<String> getErrors() {
+		return errors;
+	}
+
+	public abstract Recipe create();
+
+    public static class CraftingRecipeWrapper extends RecipeWrapper {
+		private RecipeChoice[] ingredients = new RecipeChoice[9];
+		private String group;
+		private CraftingBookCategory category;
+
+		public CraftingRecipeWrapper(NamespacedKey key, RecipeType recipeType) {
+			super(key, recipeType);
+		}
+
+		@Override
+		public Recipe create() {
+			return null;
+		}
+
+		public void setIngredients(int placement, RecipeChoice item) {
+			ingredients[placement] = item;
+		}
+
+		public void setGroup(String group) {
+			this.group = group;
+		}
+
+		public void setCategory(CraftingBookCategory category) {
+			this.category = category;
+		}
+
+		public RecipeChoice[] getIngredients() {
+			return ingredients;
+		}
+
+		public String getGroup() {
+			return this.group;
+		}
+
+		public CraftingBookCategory getCategory() {
+			return this.category;
+		}
+
+		public static class ShapedRecipeWrapper extends CraftingRecipeWrapper {
+
+			public ShapedRecipeWrapper(NamespacedKey key, RecipeType recipeType) {
+				super(key, recipeType);
+			}
+
+			@Override
+			public ShapedRecipe create() {
+				if (getResult() == null) {
+					addError("You must provide a result item when registering a recipe");
+					return null;
+				}
+				RecipeChoice[] ingredients = getIngredients();
+				if (ingredients.length == 0 || Arrays.stream(ingredients).filter(Objects::nonNull).toArray().length < 2) {
+					addError("You must have at least 1 ingredient when registering a '" + getRecipeType() + "' recipe.");
+					return null;
+				}
+				ShapedRecipe shapedRecipe = new ShapedRecipe(getKey(), getResult());
+				Character[] characters = new Character[]{'a','b','c','d','e','f','g','h','i'};
+				shapedRecipe.shape("abc","def","ghi");
+				for (int i = 0; i < ingredients.length; i++) {
+					RecipeChoice thisChoice = ingredients[i];
+					if (thisChoice != null)
+						shapedRecipe.setIngredient(characters[i], thisChoice);
+				}
+				if (getGroup() != null)
+					shapedRecipe.setGroup(getGroup());
+				if (getCategory() != null)
+					shapedRecipe.setCategory(getCategory());
+				return shapedRecipe;
+			}
+
+		}
+
+		public static class ShapelessRecipeWrapper extends CraftingRecipeWrapper {
+
+			public ShapelessRecipeWrapper(NamespacedKey key, RecipeType recipeType) {
+				super(key, recipeType);
+			}
+
+			@Override
+			public ShapelessRecipe create() {
+				if (getResult() == null) {
+					addError("You must provide a result item when registering a recipe");
+					return null;
+				}
+				RecipeChoice[] ingredients = getIngredients();
+				if (ingredients.length == 0 || Arrays.stream(ingredients).filter(Objects::nonNull).toArray().length < 2) {
```

Comment:
```suggestion
				if (ingredients.length == 0 || Arrays.stream(ingredients).filter(Objects::nonNull).count() < 2) {
```
same as above with 1 vs 2